### PR TITLE
feat(eval): treat Lua string as "blob" in writefile()

### PR DESCRIPTION
**Pull request by barrettruth**
## Problem

vim.fn.writefile() treats Lua strings as Vimscript strings instead of a "binary clean" string.

## Solution:

Treat Lua-originated strings as blob data.

Related (possibly fixes): #35589.

I looked over #31930 and addressed its review comments here; if this is the version to merge, #31930 should be closed.

### CODE REVIEW

**Comment by lewis6991** [comment]
Does this support lua strings that contain NUL characters?

**Comment by zeertzjq** [comment]
Those are converted to Blobs, which already work.

File: test/functional/vimscript/writefile_spec.lua [hunk 1]
```diff
@@ -100,6 +101,11 @@ describe('writefile()', function()
     eq('a\0', read_file(fname))
   end)
 
+  it('writes Lua strings to a file', function()
+    eq(0, exec_lua([[return vim.fn.writefile('foo\nbar', ..., 'b')]], fname))
```
Comment (justinmk) [inline test/functional/vimscript/writefile_spec.lua]: can we test writing NUL byte too?

File: src/nvim/lua/executor.c [hunk 2]
```diff
@@ -1246,12 +1246,14 @@ int nlua_call(lua_State *lstate)
   funcexe.fe_firstline = curwin->w_cursor.lnum;
   funcexe.fe_lastline = curwin->w_cursor.lnum;
   funcexe.fe_evaluate = true;
+  const sctx_T save_current_sctx = api_set_sctx(LUA_INTERNAL_CALL);
 
   TRY_WRAP(&err, {
```
Comment (justinmk) [inline src/nvim/lua/executor.c:1371]: off topic for this PR: I wonder if `TRY_WRAP` should save/restore the sctx?

File: test/functional/vimscript/writefile_spec.lua (hunk 1 above)
Comment (barrettruth) [reply test/functional/vimscript/writefile_spec.lua]: null-char included in this test now.

File: test/functional/vimscript/writefile_spec.lua [hunk 3]
```diff
@@ -99,6 +100,11 @@ describe('writefile()', function()
     eq('a\0', read_file(fname))
   end)
 
+  it('writes Lua strings to a file', function()
+    eq(0, exec_lua([[return vim.fn.writefile('foo\0bar', ..., 'b')]], fname))
```
Comment (justinmk) [inline test/functional/vimscript/writefile_spec.lua:104]: nonblocker/followup: we probably should test without 'b' too. can do a bunch of weird edge cases in this same `it()`.
